gadget/install/install.go

```diff
@@ -91,6 +96,76 @@ func saveStorageTraits(allLaidOutVols map[string]*gadget.LaidOutVolume, optsPerV
 	return nil
 }
 
+func installOneParition(part *gadget.OnDiskStructure, encryptionType secboot.EncryptionType, sectorSize quantity.Size, perfTimings timings.Measurer, observer gadget.ContentObserver) (encryptionKey keys.EncryptionKey, err error) {
+	encrypt := (encryptionType != secboot.EncryptionTypeNone)
+
+	if encrypt && roleNeedsEncryption(part.Role) {
+		var err error
+		timings.Run(perfTimings, fmt.Sprintf("make-key-set[%s]", roleOrLabelOrName(part)),
+			fmt.Sprintf("Create encryption key set for %s", roleOrLabelOrName(part)),
+			func(timings.Measurer) {
+				encryptionKey, err = keys.NewEncryptionKey()
+				if err != nil {
+					err = fmt.Errorf("cannot create encryption key: %v", err)
+				}
+			})
+		if err != nil {
+			return nil, err
+		}
+		logger.Noticef("encrypting partition device %v", part.Node)
+		var dataPart encryptedDevice
+		switch encryptionType {
+		case secboot.EncryptionTypeLUKS:
+			timings.Run(perfTimings, fmt.Sprintf("new-encrypted-device[%s]", roleOrLabelOrName(part)),
+				fmt.Sprintf("Create encryption device for %s", roleOrLabelOrName(part)),
+				func(timings.Measurer) {
+					dataPart, err = newEncryptedDeviceLUKS(part, encryptionKey, part.Label)
+				})
+			if err != nil {
+				return nil, err
+			}
+
+		case secboot.EncryptionTypeDeviceSetupHook:
+			timings.Run(perfTimings, fmt.Sprintf("new-encrypted-device-setup-hook[%s]", roleOrLabelOrName(part)),
+				fmt.Sprintf("Create encryption device for %s using device-setup-hook", roleOrLabelOrName(part)),
+				func(timings.Measurer) {
+					dataPart, err = createEncryptedDeviceWithSetupHook(part, encryptionKey, part.Name)
+				})
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// update the encrypted device node
+		part.Node = dataPart.Node()
+		logger.Noticef("encrypted device %v", part.Node)
+	}
+
+	// use the diskLayout.SectorSize here instead of lv.SectorSize, we check
+	// that if there is a sector-size specified in the gadget that it
+	// matches what is on the disk, but sometimes there may not be a sector
+	// size specified in the gadget.yaml, but we will always have the sector
+	// size from the physical disk device
```

Comment:
I see that you only moved this part and didn't write it, so we can consider it for a follow-up: this comment seems out of place, it doesn't really help me here. Maybe it should be moved to `makeFilesystem()`?

Reply:
actually it should be moved back to Run(), so I moved it back.